app/oauth.py

```diff
@@ -0,0 +1,76 @@
+import json
+import re
+
+import requests
+from oauth2_provider.models import AccessToken
+
+from django.contrib.auth.models import User as DefaultUser
+
+
+def generate_github_access_token(github_client_id, github_client_secret, github_code):
+    """
+    create an access token to github Oauth2.
+    :github_client_id: client_id from https://github.com/settings/developers
+    :github_client_secret: client_secret from https://github.com/settings/developers
+    :code: code generated by client from http://github.com/login/oauth/authorize/
+    :return: json data on user's api
+    """
+    github_response = requests.post(
+        'https://github.com/login/oauth/access_token/',
+        data=json.dumps({
+            'client_id': github_client_id,
+            'client_secret': github_client_secret,
+            'code': github_code
+        }),
+        headers={'content-type': 'application/json'}
+    )
+    assert github_response.status_code == 200, \
+        f'ERROR: {github_response.status_code}, {github_response.text}'
+    token = re.search(r"access_token=([a-zA-Z0-9]+)", github_response.content.decode('utf-8'))
+    if not token:
+        raise PermissionError(github_response)
+    return token.group(1)
+
+
+def retrieve_github_user_info(token):
+    """
+    using the access token returned by github, retrieve the user's info from the github api
+    :token: access token generated from github
+    """
+    response = requests.get(
+        'https://api.github.com/user',
+        data={'token': token},
+        headers={'Authorization': f'token {token}', 'content-type': 'application/json'}
+    )
+    return response.json()
+
+
+def convert_to_auth_token(client_id, client_secret, backend, token):
+    """
+    with the previously generated access_token, use the django-rest-framework-social-oauth2
+    endpoint `/convert-token/` to authenticate the user and return a django auth token
+    :client_id: from OauthToolkit application
+    :client_secret: from OauthToolkit application
+    :backend: social authentication backend to the user ('github', 'facebook', etc)
+    :token: access token generated from the github backend
+    :return: django auth token
+    """
+    params = {
+        'grant_type': 'convert_token',
+        'client_id': client_id,
+        'client_secret': client_secret,
+        'backend': backend,
+        'token': token,
+    }
+    response = requests.post('http://127.0.0.1:8000/auth/convert-token/', params=params)
```

Comment:
Can we make this URL `http://127.0.0.1:8000` an env variable? Or would it be used on prod as well?

Reply:
okay. for using it on prod, I don't think so. Not really sure to be honest.

Reply:
still make it an env var so that you can use the appropriate URL on prod.  localhost will work on prod.